### PR TITLE
Added the icon property.

### PR DESCRIPTION
**Pull request by ptvoinfo**
Added the icon property. The idea is to define a device's icon in custom converters and then use them in the frontend. The icon is an URL or a BASE64 encoded SVG inline image (https://css-tricks.com/lodge/svg/09-svg-data-uris/). Later, we can define icons for all devices in devices.js. The icon is optional.

### CODE REVIEW

**Comment by nurikk** [comment]
In case of devices with multiple `zigbeeModel`

`zigbeeModel: ['TS0201', 'SNTZ003']`

Can we specify device images for each modification?

**Comment by kirovilya** [comment]
it seems to me not by the zigbeeModel, but by the whiteLabel https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/devices.js#L1360

**Comment by kirovilya** [comment]
but it is not clear how to determine the current whiteLabel :)

**Comment by ptvoinfo** [comment]
@nurikk, @kirovilya  In this case, we can define the icon property as an array too.

**Comment by nurikk** [comment]
We can settle on something like substitution operators like

`icon: https://host/foo/bar/${zigbeeModel}.png`

**Comment by ptvoinfo** [comment]
> 
> 
> We can settle on something like substitution operators like
> 
> `icon: 'https://host/foo/bar/${zigbeeModel}.png'`

It will work for URL. But what about inline images?

**Comment by kirovilya** [comment]
@ptvoinfo okay, no need an array, we leave the icon for the model. if need a separate device with your own picture - will make a separate model

**Comment by kirovilya** [comment]
@ptvoinfo for inline image - use a base64 format
```data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAYAAACNMs+9AAAABmJLR0QA/wD/AP+gvaeTAAAAB3RJTUUH1ggDCwMADQ4NnwAAAFVJREFUGJWNkMEJADEIBEcbSDkXUnfSgnBVeZ8LSAjiwjyEQXSFEIcHGP9oAi+H0Bymgx9MhxbFdZE2a0s9kTZdw01ZhhYkABSwgmf1Z6r1SNyfFf4BZ+ZUExcNUQUAAAAASUVORK5CYII=```

**Comment by kirovilya** [comment]
do not use zigbeeModel as unique id - look at tuya devices

File: lib/extension/bridge.js [hunk 1]
```diff
@@ -652,14 +652,20 @@ class Bridge extends Extension {
             'bridge/groups', stringify(groups), {retain: true, qos: 0}, settings.get().mqtt.base_topic, true);
     }
 
-    getDefinitionPayload(definition) {
+    getDefinitionPayload(definition, resolved) {
         if (definition) {
+            let icon = definition.icon;
+            if (icon) {
+                icon = icon.replace('${zigbeeModel}', (resolved && resolved.hasOwnProperty('device'))?resolved.device.modelID:'');
+                icon = icon.replace('${model}', definition.model);
```
Comment (nurikk) [inline lib/extension/bridge.js:660]: Attributes should be sanitized using same rules
https://github.com/Koenkk/zigbee2mqtt.io/blob/0436c7eacee55585ca66652bce89eb3207ae51b9/docgen/utils.js#L5

Comment (nurikk) [reply lib/extension/bridge.js:660]: Some wild garbage might be in this attributes :)

File: lib/extension/bridge.js [hunk 2]
```diff
@@ -652,14 +652,26 @@ class Bridge extends Extension {
...
+    getDefinitionPayload(definition, resolved) {
         if (definition) {
+            let icon = null;
+            if (definition.hasOwnProperty('icon')) {
+                icon = definition.icon.replace('${zigbeeModel}',
+                    (resolved && resolved.hasOwnProperty('device'))?resolved.device.modelID:'');
+                icon = icon.replace('${model}', definition.model);
+
+                if (!icon.match(/^data:image\//)) {
+                    const replaceByDash = [/\?/g, /&/g, /[^\dA-Za-z\- _./:]/gi];
```
Comment (nurikk) [inline lib/extension/bridge.js:664]: This const should be moved away from function code

File: lib/extension/bridge.js [hunk 3]
```diff
@@ -652,14 +652,26 @@ class Bridge extends Extension {
             'bridge/groups', stringify(groups), {retain: true, qos: 0}, settings.get().mqtt.base_topic, true);
     }
 
-    getDefinitionPayload(definition) {
+    getDefinitionPayload(definition, resolved) {
         if (definition) {
+            let icon = null;
+            if (definition.hasOwnProperty('icon')) {
+                icon = definition.icon.replace('${zigbeeModel}',
```
Comment (nurikk) [inline lib/extension/bridge.js:659]: Looks like too much logic in one line, better splitting

File: lib/extension/bridge.js (hunk 2 above)
Comment (nurikk) [reply lib/extension/bridge.js:664]: Even worth to move to utils and import as sanitize function

File: lib/extension/bridge.js [hunk 4]
```diff
@@ -652,14 +652,28 @@ class Bridge extends Extension {
...
-    getDefinitionPayload(definition) {
+    getDefinitionPayload(definition, resolvedDevice) {
         if (definition) {
+            let icon = null;
+            if (definition.hasOwnProperty('icon')) {
+                let zigbeeModel = '';
+                if (resolvedDevice && resolvedDevice.hasOwnProperty('device')) {
+                    zigbeeModel = resolvedDevice.device.modelID;
+                }
+                icon = definition.icon.replace('${zigbeeModel}', zigbeeModel);
```
Comment (nurikk) [inline lib/extension/bridge.js:663]:
```
-icon = definition.icon.replace('${zigbeeModel}', zigbeeModel);
+icon = definition.icon.replace('${zigbeeModel}', utils.sanitizeImageLink(zigbeeModel));
```
samne for `definition.model`

File: lib/extension/bridge.js [hunk 5]
```diff
@@ -652,14 +652,28 @@ class Bridge extends Extension {
...
+                icon = icon.replace('${model}', definition.model);
+
+                if (!icon.match(/^data:image\//)) {
```
Comment (nurikk) [inline lib/extension/bridge.js:666]: this isn't necessary

Comment (ptvoinfo) [reply lib/extension/bridge.js:666]: why?

Comment (nurikk) [reply lib/extension/bridge.js:666]: you've introduced this logic to protect base64-ecoded image from corruption if you perform `icon = utils.sanitizeImageLink(icon);`, but you shouldn't sanitise a icon url, but sanitise `definition.model` and `zigbeeModel`

**Comment by ptvoinfo** [comment]
Finally, my changes passed all tests :-).

**Comment by nurikk** [comment]
LGTM

**Comment by Koenkk** [comment]
@nurikk feel free to merge this when you think it is ready

**Comment by nurikk** [comment]
@ptvoinfo hi, can you please resolve merge conflicts?

**Pull request closed without merging**